phoenix-core/src/main/java/org/apache/phoenix/schema/types/PJson.java

```diff
@@ -0,0 +1,223 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.phoenix.schema.types;
+
+import java.sql.SQLException;
+import java.sql.Types;
+import java.text.Format;
+
+import org.apache.hadoop.hbase.io.ImmutableBytesWritable;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.phoenix.schema.IllegalDataException;
+import org.apache.phoenix.schema.SortOrder;
+import org.apache.phoenix.schema.json.PhoenixJson;
+import org.apache.phoenix.util.ByteUtil;
+import org.apache.phoenix.util.StringUtil;
+
+import com.google.common.base.Preconditions;
+
+/**
+ * <p>
+ * A Phoenix data type to represent JSON. The json data type stores an exact copy of the input text,
+ * which processing functions must reparse on each execution. Because the json type stores an exact
+ * copy of the input text, it will preserve semantically-insignificant white space between tokens,
+ * as well as the order of keys within JSON objects. Also, if a JSON object within the value
+ * contains the same key more than once, all the key/value pairs are kept. It stores the data as
+ * string in single column of HBase and it has same data size limit as Phoenix's Varchar.
+ * <p>
+ * JSON data types are for storing JSON (JavaScript Object Notation) data, as specified in RFC 7159.
+ * Such data can also be stored as text, but the JSON data types have the advantage of enforcing
+ * that each stored value is valid according to the JSON rules.
+ */
+public class PJson extends PDataType<String> {
+
+    public static final PJson INSTANCE = new PJson();
+
+    PJson() {
+        super("JSON", Types.OTHER, PhoenixJson.class, null, 48);
+    }
+
+    @Override
+    public int toBytes(Object object, byte[] bytes, int offset) {
+
+        if (object == null) {
+            return 0;
+        }
+        byte[] b = toBytes(object);
+        System.arraycopy(b, 0, bytes, offset, b.length);
+        return b.length;
+
+    }
+
+    @Override
+    public byte[] toBytes(Object object) {
+        if (object == null) {
+            return ByteUtil.EMPTY_BYTE_ARRAY;
+        }
+        PhoenixJson phoenixJson = (PhoenixJson) object;
+        return PVarchar.INSTANCE.toBytes(phoenixJson.toString());
+    }
+
+    @Override
+    public Object toObject(byte[] bytes, int offset, int length,
+            @SuppressWarnings("rawtypes") PDataType actualType, SortOrder sortOrder,
+            Integer maxLength, Integer scale) {
+
+        if (!actualType.isCoercibleTo(this)) {
+            throwConstraintViolationException(actualType, this);
+        }
+        if (length == 0) {
+            return null;
+        }
```

Comment:
Here you want to call through to PVarchar to get a String, then call getPhoenixJson(String). This will take care of the maxLength check and the inversion if necessary based on sortOrder.

    String jsonStr = PVarChar.toObject(bytes, offset, length, actualType, sortOrder, maxLength, scale);
    return getPhoenixJson(jsonStr);